File: ios/Photo Editor/RNPhotoEditorViewManager.m
#import <React/RCTViewManager.h>

@interface RCT_EXTERN_MODULE(RNPhotoEditorViewManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(toolSize, CGFloat)
RCT_EXPORT_VIEW_PROPERTY(toolColor, UIColor)
RCT_EXPORT_VIEW_PROPERTY(mode, NSString)
RCT_EXPORT_VIEW_PROPERTY(needChangeConfirm, Bool)
RCT_EXPORT_VIEW_PROPERTY(onLayersUpdate, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onImageLoadError, RCTDirectEventBlock)

RCT_EXTERN_METHOD(clearAll:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(rotate: (nonnull NSNumber *)node clockwise:(BOOL)clockwise)
RCT_EXTERN_METHOD(crop:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(undo:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(redo:(nonnull NSNumber *)node)

RCT_CUSTOM_VIEW_PROPERTY(source, NSObject, RNPhotoEditorViewManager)
{
    [self performSelector:@selector(setSource:) withObject:@{@"view":view, @"json":json}];

}
@end

File: ios/RNPhotoEditorViewManager.swift
import Foundation
import React

@objc(RNPhotoEditorViewManager)
class RNPhotoEditorViewManager: RCTViewManager {
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }

    override func view() -> UIView! {
        return RNPhotoEditorView()
    }

    @objc(setSource:)
    public func setSource(obj:Dictionary<String, Any>) {
        guard let currentView = obj["view"] as? RNPhotoEditorView,
              let src = obj["json"] as? NSDictionary else {
                  return
              }

        currentView.setSource(source: src)
    }

    @objc func clearAll(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! RNPhotoEditorView
            component.clearAll()
        }
    }

    @objc func rotate(_ node: NSNumber, clockwise: Bool) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! RNPhotoEditorView
            component.rotate(clockwise:clockwise)
        }
    }
    
    @objc func crop(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! RNPhotoEditorView
            component.crop()
        }
    }

    @objc func reload(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! RNPhotoEditorView
            component.reload()
        }
    }

    @objc func processPhoto(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! RNPhotoEditorView
            component.processPhoto()
        }
    }

    @objc func undo(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! RNPhotoEditorView
            component.undo()
        }
    }
    @objc func redo(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! RNPhotoEditorView
            component.redo()
        }
    }
}
